packages/devtools/lib/src/timeline/event_details.dart

```diff
@@ -266,71 +321,116 @@ class _UiEventDetails extends CoreElement {
       data: cpuProfileData,
-      flameChartWidth: flameChart.element.clientWidth,
-      flameChartHeight:
-          cpuProfileData.cpuProfileRoot.depth * rowHeightWithPadding,
+      flameChartWidth: element.clientWidth,
+      flameChartHeight: math.max(
+        // Subtract [rowHeightWithPadding] to account for timeline at the top of
+        // the flame chart.
+        element.clientHeight - rowHeightWithPadding,
+        // Add 1 to account for a row of padding at the bottom of the chart.
+        (cpuProfileData.cpuProfileRoot.depth + 1) * rowHeightWithPadding,
+      ),
     );
 
-    flameChartCanvas.onStackFrameSelected.listen((CpuStackFrame stackFrame) {
-      _updateStackFrameDetails(stackFrame);
+    // Add the last known value to the stream so that the canvas reflects the
+    // persisted checkbox state.
+    _collapseNativeSamplesController.add(lastCollapseNativeSamplesValue);
+
+    canvas.onStackFrameSelected.listen((CpuStackFrame stackFrame) {
+      stackFrameDetails.text = stackFrame.toString();
     });
 
-    flameChart.add(flameChartCanvas.element);
+    add(canvas.element);
   }
 
-  void _updateFlameChartForError(CoreElement errorDiv) {
-    flameChart.add(errorDiv);
-    showingFlameChartError = true;
+  void _updateChartWithError(String message) {
+    showingError = true;
+    error.text = message;
+    error.attribute('hidden', false);
     stackFrameDetails.attribute('hidden', true);
   }
 
   Future<void> update(TimelineEvent event) async {
-    if (event == this.event) {
-      return;
-    }
-
     reset();
     this.event = event;
 
-    final Spinner spinner = Spinner()..clazz('cpu-profile-spinner');
-    add(spinner);
+    // Update the canvas if the flame chart is visible. Otherwise, mark the
+    // canvas as needing a rebuild.
+    if (!isHidden) {
+      final Spinner spinner = Spinner()..clazz('cpu-profile-spinner');
+      add(spinner);
+
+      try {
+        await _drawFlameChart();
+
+        if (!showingError) {
+          stackFrameDetails.text = stackFrameDetailsDefaultText;
+          stackFrameDetails.attribute('hidden', false);
+        }
+      } catch (e) {
+        _updateChartWithError('Error retrieving CPU profile: ${e.toString()}');
+      }
 
-    try {
-      // TODO(kenzie): add a timeout here so we don't appear to have an
-      // infinite spinner.
-      await _drawFlameChart();
-    } catch (e) {
-      _updateFlameChartForError(div(
-          text: 'Error retrieving CPU profile: ${e.toString()}', c: 'message'));
+      spinner.element.remove();
+    } else {
+      canvasNeedsRebuild = true;
     }
+  }
 
-    spinner.element.remove();
+  void updateForContainerResize() {
+    if (canvas == null) {
+      return;
+    }
 
-    stackFrameDetails.text = stackFrameDetailsDefaultText;
+    // Only update the canvas if the flame chart is visible and has data.
+    // Otherwise, mark the canvas as needing a rebuild.
+    if (!isHidden && cpuProfileData != null) {
```

Comment:
add a comment on why the content size needs to change if the container size changes.

Reply:
done.